### ISSUE TITLE
Executing HelloFX in Ubuntu 20-04 (focus) getting RuntimeExceptions

### ISSUE THREAD

**Issue by acgrimes** [gluonhq/substrate#611]
My environment is:
Apache Maven 3.6.3
Maven home: /usr/share/maven
Java version: 11.0.6, vendor: Oracle Corporation, runtime: /home/linuxlp/opt/graalvm/graalvm-svm-linux-20.1.0-ea+28
Default locale: en_US, platform encoding: UTF-8
OS name: "linux", version: "5.4.0-31-generic", arch: "amd64", family: "unix"

The Issue:
Our application deployed to the desktop in Ubuntu 19-10. The gluon-samples also deployed sucessfully to the desktop with Ubuntu 19-10. After upgrading to Ubuntu 20-04 none of the applications are deploying to the desktop now. However, on a macBook Pro all applications continue to deploy to the desktop and an Iphone 7 Plus. 

Running HelloFX on Ubuntu 20-04 "mvn client:run" fails silently with:

linuxlp@dd-ThinkPad-P71:~/codeExecute/HelloFX$ mvn client:run
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by com.google.inject.internal.cglib.core.$ReflectUtils$1 (file:/usr/share/maven/lib/guice.jar) to method java.lang.ClassLoader.defineClass(java.lang.String,byte[],int,int,java.security.ProtectionDomain)
WARNING: Please consider reporting this to the maintainers of com.google.inject.internal.cglib.core.$ReflectUtils$1
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
[INFO] Scanning for projects...
[INFO] 
[INFO] ---------------------------< hello:hellofx >----------------------------
[INFO] Building hellofx 1.0-SNAPSHOT
[INFO] --------------------------------[ jar ]---------------------------------
[INFO] 
[INFO] --- client-maven-plugin:0.1.23:run (default-cli) @ hellofx ---
[Thu May 21 11:58:31 EDT 2020][INFO] ==================== RUN TASK ====================
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  0.756 s
[INFO] Finished at: 2020-05-21T11:58:32-04:00
[INFO] ------------------------------------------------------------------------

Running HelloFX directly as "./hellofx" the following runtime exceptions occur:

linuxlp@dd-ThinkPad-P71:~/codeExecute/HelloFX/target/client/x86_64-linux$ ./hellofx
Graphics Device initialization failed for :  es2, sw
Error initializing QuantumRenderer: no suitable pipeline found
java.lang.RuntimeException: java.lang.RuntimeException: Error initializing QuantumRenderer: no suitable pipeline found
	at com.sun.javafx.tk.quantum.QuantumRenderer.getInstance(QuantumRenderer.java:280)
	at com.sun.javafx.tk.quantum.QuantumToolkit.init(QuantumToolkit.java:244)
	at com.sun.javafx.tk.Toolkit.getToolkit(Toolkit.java:260)
	at com.sun.javafx.application.PlatformImpl.startup(PlatformImpl.java:267)
	at com.sun.javafx.application.PlatformImpl.startup(PlatformImpl.java:158)
	at com.sun.javafx.application.LauncherImpl.startToolkit(LauncherImpl.java:658)
	at com.sun.javafx.application.LauncherImpl.launchApplication1(LauncherImpl.java:678)
	at com.sun.javafx.application.LauncherImpl.lambda$launchApplication$2(LauncherImpl.java:195)
	at java.lang.Thread.run(Thread.java:834)
	at com.oracle.svm.core.thread.JavaThreads.threadStartRoutine(JavaThreads.java:517)
	at com.oracle.svm.core.posix.thread.PosixJavaThreads.pthreadStartRoutine(PosixJavaThreads.java:193)
Caused by: java.lang.RuntimeException: Error initializing QuantumRenderer: no suitable pipeline found
	at com.sun.javafx.tk.quantum.QuantumRenderer$PipelineRunnable.init(QuantumRenderer.java:94)
	at com.sun.javafx.tk.quantum.QuantumRenderer$PipelineRunnable.run(QuantumRenderer.java:124)
	... 3 more
Exception in thread "main" java.lang.RuntimeException: No toolkit found
	at com.sun.javafx.tk.Toolkit.getToolkit(Toolkit.java:272)
	at com.sun.javafx.application.PlatformImpl.startup(PlatformImpl.java:267)
	at com.sun.javafx.application.PlatformImpl.startup(PlatformImpl.java:158)
	at com.sun.javafx.application.LauncherImpl.startToolkit(LauncherImpl.java:658)
	at com.sun.javafx.application.LauncherImpl.launchApplication1(LauncherImpl.java:678)
	at com.sun.javafx.application.LauncherImpl.lambda$launchApplication$2(LauncherImpl.java:195)
	at java.lang.Thread.run(Thread.java:834)
	at com.oracle.svm.core.thread.JavaThreads.threadStartRoutine(JavaThreads.java:517)
	at com.oracle.svm.core.posix.thread.PosixJavaThreads.pthreadStartRoutine(PosixJavaThreads.java:193)

**Comment by jperedadnr** [comment]
I had Ubuntu 19.10, and I've just upgraded it to 20.04. In both cases, HelloFX runs fine for me.

@acgrimes Can you edit your post and add the content of the files under `~/codeExecute/HelloFX/target/client/x86_64-linux/gvm/log`?

**Comment by acgrimes** [comment]
Adding files:
[client-debug0.log](https://github.com/gluonhq/substrate/files/4663785/client-debug0.log)
[process-compile-1590070332188.log](https://github.com/gluonhq/substrate/files/4663786/process-compile-1590070332188.log)
[process-link-1590070348065.log](https://github.com/gluonhq/substrate/files/4663787/process-link-1590070348065.log)

**Comment by jperedadnr** [comment]
Can you add this to the plugin's configuration in your pom:
```
<nativeImageArgs>
      <arg>-Dprism.verbose=true</arg>
</nativeImageArgs>
```
then update it to version `0.1.24` and try again?

**Comment by acgrimes** [comment]
updated plugin version:
      <client.plugin.version>0.1.24</client.plugin.version>

added to configuration:
            <plugin>
                <groupId>com.gluonhq</groupId>
                <artifactId>client-maven-plugin</artifactId>
                <version>${client.plugin.version}</version>
                <configuration>
                	<nativeImageArgs>
      					<arg>-Dprism.verbose=true</arg>
					</nativeImageArgs>
                    
                    
                    <mainClass>${mainClassName}</mainClass>
                </configuration>
            </plugin>

Appears to get the same exceptions.
[client-debug0.log](https://github.com/gluonhq/substrate/files/4664021/client-debug0.log)
[client-debug01.log](https://github.com/gluonhq/substrate/files/4664026/client-debug01.log)

**Comment by jperedadnr** [comment]
Ok, thx. Can you run:
 `~/codeExecute/HelloFX/target/client/x86_64-linux/HelloFX -Dprism.verbose=true`
and post the console output?

**Comment by jperedadnr** [comment]
See this line:
```
Failed Graphics Hardware Qualifier check
```
Looks like your GPU is somehow blacklisted.

You can add:
`<enableSWRendering>true</enableSWRendering>` to your plugin configuration, build and run again:

`~/codeExecute/HelloFX/target/client/x86_64-linux$ ./hellofx -Dprism.verbose=true`

Can you post the console output?

**Comment by jperedadnr** [comment]
And once built, please also run (and post result):
 
`~/codeExecute/HelloFX/target/client/x86_64-linux$ ./hellofx -Dprism.verbose=true -Dprism.forceGPU=true`

**Comment by acgrimes** [comment]
That worked, the image deployed to the desktop. Thanks for taking the time to make this work for us. We need to find out what the issue with our GPU is.

linuxlp@dd-ThinkPad-P71:~/codeExecute/HelloFX/target/client/x86_64-linux$ ./hellofx -Dprism.verbose=true -Dprism.forceGPU=true
Prism pipeline init order: es2 sw 
Using Double Precision Marlin Rasterizer
Using dirty region optimizations
Not using texture mask for primitives
Not forcing power of 2 sizes for textures
Using hardware CLAMP_TO_ZERO mode
Opting in for HiDPI pixel scaling
Prism pipeline name = com.sun.prism.es2.ES2Pipeline
Loading ES2 native library ... prism_es2
	succeeded.
GLFactory using com.sun.prism.es2.X11GLFactory
(X) Got class = class com.sun.prism.es2.ES2Pipeline
Initialized prism pipeline: com.sun.prism.es2.ES2Pipeline
Maximum supported texture size: 16384
Maximum texture size clamped to 4096
Non power of two texture support = true
Maximum number of vertex attributes = 16
Maximum number of uniform vertex components = 16384
Maximum number of uniform fragment components = 16384
Maximum number of varying components = 128
Maximum number of texture units usable in a vertex shader = 32
Maximum number of texture units usable in a fragment shader = 32
Graphics Vendor: Intel
       Renderer: Mesa Intel(R) HD Graphics 630 (KBL GT2)
        Version: 4.6 (Compatibility Profile) Mesa 20.0.4
 vsync: true vpipe: true
ES2ResourceFactory: Prism - createStockShader: FillPgram_Color.frag
ES2ResourceFactory: Prism - createStockShader: Solid_TextureRGB.frag
ES2ResourceFactory: Prism - createStockShader: Solid_TextureFirstPassLCD.frag
ES2ResourceFactory: Prism - createStockShader: Solid_TextureSecondPassLCD.frag

**Comment by jperedadnr** [comment]
Ok, good to know.

For some reason, this is the check that fails:
https://github.com/openjdk/jfx/blob/ce5ec2ca156e9e276d393997f57dc0483ffc2ccc/modules/javafx.graphics/src/main/java/com/sun/prism/es2/GLFactory.java#L184

With a little bit of "simple" inspection, your graphics vendor "Intel" doesn't match "Intel Open Source Technology Center":

https://github.com/openjdk/jfx/blob/ce5ec2ca156e9e276d393997f57dc0483ffc2ccc/modules/javafx.graphics/src/main/java/com/sun/prism/es2/X11GLFactory.java#L45

I wonder:
- Does `mvn javafx:run` works for you?ç

- Did Ubuntu 19.10 show "Intel Open Source Technology Center"? (I guess that won't be easy to find out now, but seems that could be the case as in https://stackoverflow.com/a/60051826/3956070). But that might explain the issue?

It seems there are many related issues about Intel drivers with Ubuntu 20.04 (for instance https://bugs.launchpad.net/ubuntu/+source/xserver-xorg-video-intel/+bug/1864778).

**Comment by bgmf** [comment]
Looks, like I'm facing the same issue. Forcing the SW renderer "solves" the issue.

System: Dell XPS 15 9550, i7-6700HQ, with integrated Intel HD Graphics 530 and discrete NVIDIA GeForce GTX 960M. The selected HW driver is the proprietary one, nVidia settings tool reports the discrete is in use.

```
daniel@ghua-xps15:~$ glxinfo | grep vendor
server glx vendor string: SGI
client glx vendor string: Mesa Project and SGI
OpenGL vendor string: Intel
daniel@ghua-xps15:~$ glxinfo | grep renderer
    GLX_MESA_multithread_makecurrent, GLX_MESA_query_renderer, 
    GLX_MESA_query_renderer, GLX_MESA_swap_control, GLX_OML_swap_method, 
Extended renderer info (GLX_MESA_query_renderer):
OpenGL renderer string: Mesa Intel(R) HD Graphics 530 (SKL GT2)
```
I guess this suggests, I'm still using the integrated GPU which is named  `Intel`, correct?

This is the output, when I add `-Dprism.verbose=true -Dprism.forceGPU=true`

```
Graphics Vendor: Intel
       Renderer: Mesa Intel(R) HD Graphics 530 (SKL GT2)
        Version: 4.6 (Compatibility Profile) Mesa 20.0.4
```

By the way, `lspci -k | grep -EA3 'VGA|3D|Display'` reports
```
00:02.0 VGA compatible controller: Intel Corporation HD Graphics 530 (rev 06)
	DeviceName:  Onboard IGD
	Subsystem: Dell XPS 15 9550
	Kernel driver in use: i915
--
01:00.0 3D controller: NVIDIA Corporation GM107M [GeForce GTX 960M] (rev a2)
	Subsystem: Dell XPS 15 9550
	Kernel modules: nvidiafb, nouveau, nvidia_drm, nvidia
```
Also: `prime-select query` reports `nvidia`...

---

So basically: I've installed the proprietary driver, in theory my system is using the discrete GPU, but it still is using Intel (with a "wrong"/not-anticipated name) to render OpenGL...

**Comment by jperedadnr** [comment]
Thanks for reporting. As I mentioned before:

> your graphics vendor "Intel" doesn't match "Intel Open Source Technology Center":

(OpenJFX [line](https://github.com/openjdk/jfx/blob/ce5ec2ca156e9e276d393997f57dc0483ffc2ccc/modules/javafx.graphics/src/main/java/com/sun/prism/es2/X11GLFactory.java#L45))

However, I wonder if JavaFX runs fine on hotspot (`mvn javafx:run`). 
The check is the same, it shouldn't fail only with native-image (`mvn client:build client:run`).

**Comment by mipastgt** [comment]
No, it does not. I recently bought a new Intel NUC in order to be able to run Linux for building JavaFX apps for Android.

When I just run the application with hotspot on the host it tells me that the conditional feature 3D is not supported. However, if, as I have just learned, run it with forceGPU true, it seems to work perfectly. The output of prism verbose is the following:
```
Prism pipeline init order: es2 sw 
Using Double Precision Marlin Rasterizer
Using dirty region optimizations
Not using texture mask for primitives
Not forcing power of 2 sizes for textures
Using hardware CLAMP_TO_ZERO mode
Opting in for HiDPI pixel scaling
Prism pipeline name = com.sun.prism.es2.ES2Pipeline
Loading ES2 native library ... prism_es2
	succeeded.
GLFactory using com.sun.prism.es2.X11GLFactory
(X) Got class = class com.sun.prism.es2.ES2Pipeline
Initialized prism pipeline: com.sun.prism.es2.ES2Pipeline
Maximum supported texture size: 16384
Maximum texture size clamped to 4096
Non power of two texture support = true
Maximum number of vertex attributes = 16
Maximum number of uniform vertex components = 16384
Maximum number of uniform fragment components = 16384
Maximum number of varying components = 128
Maximum number of texture units usable in a vertex shader = 32
Maximum number of texture units usable in a fragment shader = 32
Graphics Vendor: Intel
       Renderer: Mesa Intel(R) UHD Graphics (CML GT2)
        Version: 4.6 (Compatibility Profile) Mesa 20.0.4
 vsync: true vpipe: true
ES2ResourceFactory: Prism - createStockShader: FillPgram_Color.frag
ES2ResourceFactory: Prism - createStockShader: Solid_TextureFirstPassLCD.frag
ES2ResourceFactory: Prism - createStockShader: Solid_TextureSecondPassLCD.frag
ES2ResourceFactory: Prism - createStockShader: Solid_TextureRGB.frag
new alphas with length = 4096
ES2ResourceFactory: Prism - createStockShader: Texture_Color.frag
new alphas with length = 8192
ES2ResourceFactory: Prism - createStockShader: FillPgram_LinearGradient_PAD.frag
ES2ResourceFactory: Prism - createStockShader: Mask_TextureSuper.frag
PPSRenderer: scenario.effect - createShader: LinearConvolveShadow_4
ES2ResourceFactory: Prism - createStockShader: DrawPgram_Color.frag
ES2ResourceFactory: Prism - createStockShader: Texture_LinearGradient_PAD.frag
PPSRenderer: scenario.effect - createShader: LinearConvolveShadow_20
new alphas with length = 212992
```
So, from what I have learned in this thread, it would just be necessary to add "Intel" to the list of supported graphics vendors.

**Comment by jperedadnr** [comment]
@mipastgt Thanks for reporting, I wonder why nobody has filed a bug on this yet?

**Comment by mipastgt** [comment]
I wanted to report that myself already for some weeks but as Linux is not my primary system I tried to concentrate on my (FX)Android problems first.

**Comment by mipastgt** [comment]
I just filed the bug in JBS: https://bugs.openjdk.java.net/browse/JDK-8246204

**Comment by mipastgt** [comment]
And the pull request for the change is now also on the way: https://github.com/openjdk/jfx/pull/243

**Comment by mipastgt** [comment]
... and is now integrated into the master repository. So with the next EA release of JavaFX this should work again.

**Comment by jperedadnr** [comment]
Should be fixed with Substrate 0.0.26, Client plugin 0.1.28.
